torch/csrc/autograd/engine.cpp

```diff
@@ -204,7 +204,9 @@ static variable_list call_function(FunctionTask& task) {
     auto& callback = it_p.first->second;
     if (!callback(&fn, inputs)) return variable_list(fn.next_functions.size());
   }
-
+  if(!task.base->keep_graph) {
```

Comment:
The interpreter needs to know _before_ it executes that it should be releasing variables. If you don't like this way, I am open to others, but it does need to know ahead of time for efficient execution and small memory footprint.

Reply:
This seems fine to me.